sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodOverloadOptionalParameter.cs

```diff
@@ -147,18 +148,28 @@ private static bool IsMethodHidingOriginal(IMethodSymbol candidateHidingMethod,
         {
             for (int i = 0; i < candidateHidingMethod.Parameters.Length; i++)
             {
-                if (!candidateHidingMethod.Parameters[i].Type.Equals(method.Parameters[i].Type))
+                var candidateParam = candidateHidingMethod.Parameters[i];
+                var methodParam = method.Parameters[i];
+
+                if (!Equals(candidateParam.Type, methodParam.Type) &&
+                    !AreTypeParameters(candidateParam, methodParam) &&
+                    !Equals(candidateParam.Type.OriginalDefinition, methodParam.Type.OriginalDefinition))
                 {
                     return false;
                 }
 
-                if (candidateHidingMethod.Parameters[i].IsOptional != method.Parameters[i].IsOptional)
+                if (candidateParam.IsOptional != methodParam.IsOptional)

```

Comment:
That will still be an if and not a return, as we do not return when condition is true. It would look like below. The original version has two ifs which IMHO is much readable than this:

if ((!Equals(candidateParam.Type, methodParam.Type) &&
    !candidateParam.Type.Is(TypeKind.TypeParameter) &&
    methodParam.Type.Is(TypeKind.TypeParameter) &&
    !Equals(candidateParam.Type.OriginalDefinition, methodParam.Type.OriginalDefinition)) ||
    (candidateParam.IsOptional != methodParam.IsOptional))
{
    return false;
}

I'm inclined to leave it as it is - perhaps give AreTypeParameters method a better name (suggestions welcome). Grab me for a chat if you have a strong preference. Otherwise, I'm going to merge it Monday-ish (when RC is done).